openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb

```diff
@@ -0,0 +1,744 @@
+    "            nn.Linear(512, 256),\n",
+    "            nn.LeakyReLU(0.2, inplace=True),\n",
+    "            nn.Linear(256, 1),\n",
+    "            nn.Sigmoid(),\n",
+    "        )\n",
+    "\n",
+    "    def forward(self, img):\n",
+    "        img_flat = img.view(img.size(0), -1)\n",
+    "        img_flat = img_flat.float()\n",
+    "        validity = self.model(img_flat)\n",
+    "\n",
+    "        return validity"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6cc92e98",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class GAN(LightningModule):\n",
+    "    def __init__(\n",
+    "        self,\n",
+    "        channels,\n",
+    "        width,\n",
+    "        height,\n",
+    "        train_disc_only,\n",
+    "        train_gen_only,\n",
+    "        latent_dim: int = 100,\n",
+    "        lr: float = 0.0002,\n",
+    "        b1: float = 0.5,\n",
+    "        b2: float = 0.999,\n",
+    "        batch_size: int = 256,\n",
+    "        **kwargs\n",
+    "    ):\n",
+    "        super().__init__()\n",
+    "        self.save_hyperparameters()\n",
+    "\n",
+    "        data_shape = (channels, width, height)\n",
+    "        self.generator = Generator(\n",
+    "            latent_dim=self.hparams.latent_dim, img_shape=data_shape\n",
+    "        )\n",
+    "        self.discriminator = Discriminator(img_shape=data_shape)\n",
+    "\n",
+    "        self.validation_z = torch.randn(8, self.hparams.latent_dim)\n",
+    "        self.example_input_array = torch.zeros(2, self.hparams.latent_dim)\n",
+    "        self.train_disc_only = train_disc_only\n",
+    "        self.train_gen_only = train_gen_only\n",
+    "\n",
+    "    def forward(self, z):\n",
+    "        return self.generator(z)\n",
+    "\n",
+    "    def adversarial_loss(self, y_hat, y):\n",
+    "        return F.binary_cross_entropy(y_hat, y)\n",
+    "\n",
+    "    def training_step(self, batch, batch_idx, optimizer_idx):\n",
+    "        imgs, _ = batch\n",
+    "\n",
+    "        # sample noise\n",
+    "        z = torch.randn(imgs.shape[0], self.hparams.latent_dim)\n",
+    "        z = z.type_as(imgs)\n",
+    "\n",
+    "        if optimizer_idx == 0 and self.train_disc_only == 0:\n",
+    "            return self.train_generator(imgs, z, logdir=\"../generated_images\")\n",
+    "\n",
+    "        elif optimizer_idx == 1 and self.train_gen_only == 0:\n",
+    "            return self.train_discriminator(imgs, z)\n",
+    "\n",
+    "    def train_generator(self, imgs, z, logdir=\"../generated_images\"):\n",
+    "        self.generated_imgs = self(z)\n",
+    "        sample_imgs = self.generated_imgs[:6]\n",
+    "        sample_imgs = np.reshape(sample_imgs.detach().cpu().numpy(), (6, 28, 28, 1))\n",
+    "\n",
+    "        if not os.path.exists(logdir):\n",
+    "            os.mkdir(logdir)\n",
+    "        tf.autograph.set_verbosity(0)\n",
```

Comment:
Can we avoid usage of TensorFlow at all? Looks like we are downloading TensorFlow for only 3 lines of code :
```
tf.autograph.set_verbosity(0)
        file_writer = tf.summary.create_file_writer(logdir)

        with file_writer.as_default():
            tf.summary.image(
                "Generated images",
                np.reshape(sample_imgs, (-1, 28, 28, 1)),
                max_outputs=10,
                step=0,
            )
```
I am pretty sure we can avoid doing that.

Reply:
@alexey-gruzdev we are using tensorboard to show generated images. Do we want to view images directly in the notebook?